falcon/asgi/app.py

```diff
@@ -777,36 +890,66 @@ def _prepare_middleware(self, middleware=None, independent_middleware=False):
     async def _http_status_handler(self, req, resp, status, params):
         self._compose_status_response(req, resp, status)
 
-    async def _http_error_handler(self, req, resp, error, params):
-        self._compose_error_response(req, resp, error)
+    async def _http_error_handler(self, req, resp, error, params, ws=None):
+        if resp:
+            self._compose_error_response(req, resp, error)
+
+        if ws:
+            falcon._logger.warning(
+                '[FALCON] WebSocket handshake rejected due to '
+                f'raised HTTP error: {error}'
+            )
+
+            code = 3000 + falcon.util.http_status_to_code(error.status)
+            await ws.close(code)
+
+    async def _python_error_handler(self, req, resp, error, params, ws=None):
+        falcon._logger.error('[FALCON] Unhandled exception in ASGI app', exc_info=error)
+
+        if resp:
+            self._compose_error_response(req, resp, falcon.HTTPInternalServerError())
 
-    async def _python_error_handler(self, req, resp, error, params):
-        falcon._logger.error('Unhandled exception in ASGI app', exc_info=error)
-        self._compose_error_response(req, resp, falcon.HTTPInternalServerError())
+        if ws:
+            await self._ws_cleanup_on_error(ws)
 
-    async def _handle_exception(self, req, resp, ex, params):
+    async def _ws_disconnected_error_handler(self, req, resp, error, params, ws):
+        falcon._logger.warning(f'[FALCON] WebSocket client disconnected with code {error.code}')
```

Comment:
The main advantage of using `%s` is that the formatting is only done if it's actually needed. But I agree that in this case it would not change much.

Reply:
This will probably be a common occurrence, esp. if the app developer is lazy and does not catch ``WebSocketDisconnected`` within the responder. With that in mind, I switched to `debug()` (to avoid spamming the log) and `%` formatting (even though perf isn't as critical here, might as well).